feat: qlever and graphdb spatial query support and inclusion of extent in responses

recalcitrantsupplant opened the pull request:
This PR:  

- adds support for geospatial query formats used in Qlever and GraphDB, for both CQL filters and the bbox query param. Qlever intends to support GeoSPARQL in future, at which point the Qlever specific functionality can be removed.
- now includes the extent for geo:FeatureCollections in the `features/collections` endpoint responses, where a `geo:hasBoundingBox` triple is present for the Feature Collection.  

Minor changes:
- add support for `startindex` query parameter used by some ESRI products like arcmap to support pagination. NB they should suport limit (supported) & offset (unsupported) as per OGC Features spec.
- add support for `offset` query parameter (used by OGC Features API), NB this is in addition to `page` which Prez also supports
- allow and ignore "f" (mediatype) as a parameter, sent in older versions of arcgis.

The generated query format can be specified using the configuration parameter `spatial_query_format` e.g. by using an environment variable. The allowed values are: "geosparql", "qlever", "graphdb".

Other changes: 

**Summary of Support & CRS Handling:**

The previous separated bbox query param filtering and CQL spatial function filtering have been merged in the `bbox_filter.py` module which has been renamed to `spatial_filter.py`. The bbox query param code and CQL spatial filtering now import/use this module.

- __GraphDB:__

  - Utilizes standard GeoSPARQL property functions (e.g., `geo:sfIntersects`).
  - __CRS Injection:__ If a CRS is specified in the CQL query, its URI is prefixed directly to the WKT string within the `RDFLiteral` (e.g., `"<CRS_URI> WKT_STRING"^^geo:wktLiteral`). This occurs in `prez/services/query_generation/cql.py` within the `_handle_spatial` method.

- __QLever:__

  - Employs a custom `SERVICE <https://qlever.cs.uni-freiburg.de/spatialSearch/>` block.
  - __CRS Handling:__ QLever expects a plain WKT string (no CRS prefix, a `geo:wktLiteral` datatype is included however qlever ignores this). If a CRS is specified for the query, the input geometry coordinates must be transformed to this CRS *before* the plain WKT string is generated. The plain WKT is then passed to QLever, via a `VALUES` clause.

__Example Query String Snippets:__

Consider a CQL query: `s_intersects(geometry, POINT(150 -30))` with CRS `http://www.opengis.net/def/crs/OGC/1.3/CRS84`.

1. __GraphDB SPARQL Snippet (Illustrative):__

   ```sparql
   ?focus_node geo:hasGeometry ?geom_bnode .
   ?geom_bnode geo:asWKT ?geom_var .
   ?geom_bnode geo:sfIntersects "<http://www.opengis.net/def/crs/OGC/1.3/CRS84> POINT (150.0 -30.0)"^^geo:wktLiteral .
   ```

NB Prez currently uses the `?geom_bnode` as the subject of the spatial function; `?focus_node` could also be used. (The triples with predicates `geo:hasGeometry` and `geo:asWKT` are injected to match triples in the construct clause.)

2. __QLever SPARQL Snippet (Illustrative):__

   ```sparql
   ?focus_node geo:hasGeometry ?geom_bnode .
   ?geom_bnode geo:asWKT ?geom_var .

   VALUES ?wkt_input_for_qlever { "POINT (150.0 -30.0)"^^geo:wktLiteral }
   SERVICE <https://qlever.cs.uni-freiburg.de/spatialSearch/> {
       _:config qlss:algorithm qlss:libspatialjoin ;
                 qlss:left ?wkt_input_for_qlever ;
                 qlss:right ?geom_var ;
                 qlss:payload ?focus_node ;
                 qlss:joinType qlss:intersects .
       {
         ?focus_node geo:hasGeometry ?geom_bnode .
         ?geom_bnode geo:asWKT ?geom_var .
       }
   }
   ```

ashleysommer commented on prez/services/query_generation/spatial_filter.py:
```diff
@@ -0,0 +1,344 @@
+
+def get_wkt_from_coords(coordinates, geom_type: str):
+    max_decimals = find_max_decimals([(geom_type, coordinates, None)])
+    return dumps({"type": geom_type, "coordinates": coordinates}, max_decimals)
```
If you include the meta object in the geojson-like dict:
`"meta": {"srid":7844}`
Then the wkt `dumps` routine will automatically reverse the axis-order of coordinates if required in the WKT string.

In GeoSPARQL WKT, it follows the same rules as eWKT, that is, the axis order is long-lat when CRS is OGC CRS84 (or if CRS tag is omitted), but axis order is lat-long when CRS is an epsg code.

The OGC-Features-API states extents are always entered in long-lat pairs (same as all OGC standards, including CRS84), so if the user specifies a CRS like "7844" in their query, then when dumping those out coords to a WKT string, the axis order _does_ need to be reversed.

_Note_: When you include the `"meta":{"srid":nnn}` obj in `wkt.dumps()` the result will be an _eWKT_ string, so it will look like `SRID=7844; POLYGON((y, x), (y, x))`
So you will need to replace that SRID part with the GeoSPARQL-compatible CRS IRI `<http://www.opengis.net/def/crs/EPSG/0/7844> POLYGON (())`

ashleysommer replied:
Is the supplied CRS in the CQL query always a URI? Can it be just a number too?

The kinds of CRS-IRIs that my extension to geomet handles are these:
```
# The default.
<http://www.opengis.net/def/crs/OGC/1.3/CRS84>
I don't know if there are any other individuals in the /OGC/1.3/ space, but CRS84 is the only one I know of.

# The OpenGIS EPSG defs:
<http://www.opengis.net/def/crs/EPSG/0/7844>

# Shortcut for above, I actually don't know if these ones are recognised by GraphDB
<https://epsg.io/7844>

# URNs:
<ogc:def:crs:epsg:7844>
```
The logic I use is:
1) split off the last part based on last colon (if its a URI) or the last forward-slash.
2) if the last part is "CRS84" (upper) or "crs84" (lower) then don't do any special handling
3) If the last part is all numbers, treat it as an EPSG code.
4) It its something else, treat it like its CRS84.

recalcitrantsupplant replied:
It must be a URI: https://docs.ogc.org/is/18-058/18-058.html#_discovery

So I think I should just test with your above suggested change `"meta": {"srid":7844}` with the full URI not the short code.

ashleysommer commented:
@recalcitrantsupplant 
This is a lot of work, it looks great.

> NB Prez currently uses the ?geom_bnode as the subject of the spatial function; ?focus_node could also be used.

True, but I think the ?geom node does make more sense, I prefer that. And it allows us to use schema:spatial instead of geo:hasGeometry too like:
```turtle
ex:myFeature a geo:Feature ;
    schema:spatial [
        a geo:Geometry ;
        geo:asWKT "<POINT>"^^geo:wktLiteral .
    ] ;
.
```

I was wondering how you were loading the WKT string with a GeoSPARQL-aware WKT parser, then I noticed you're using the (heavily-modified) version of geomet that is included in the rdf2geojson package. Thats the right solution, it is the only python lib that can parse a GeoSPARQL-comaptible WKT string (that has a CRS IRI at the start), and it also converts the axis-order of the WKT string to the GeoJSON axis order if required (depending on the WKT's EPSG code).

lalewis1 commented:
Still investigating, but the new optional offset param is still giving a 400 response when included

recalcitrantsupplant commented:
> Still investigating, but the new optional offset param is still giving a 400 response when included.
> 
> it works for `/catalogs?offset=2`
> 
> but 400 on `/features/collections?offset=2`

I implemented the functionality but forgot to allow it as a query param. Fixed in https://github.com/RDFLib/prez/pull/373/commits/9be8bd428191faa637b3cb845bc77fb612e37f0d

lalewis1 commented:
ah awesome, thanks

recalcitrantsupplant commented on prez/services/query_generation/spatial_filter.py:
```diff
@@ -0,0 +1,358 @@
+    if filter_crs == "http://www.opengis.net/def/crs/OGC/1.3/CRS84":
+        srid = "CRS84"
+    else:
+        srid = filter_crs  # axes will be reorded by geomet if anything other than CRS84
+    wkt_with_srid = dumps({"type": geom_type, "coordinates": coordinates, "meta": {"srid": srid}}, max_decimals)
```
Yeah I might try clean this up, it works now but could regress later if geomet/your branch is updated. @rob-metalinkage @avillar is there an OGC reference data mapping of CRS URIs to SRIDS? e.g. from `http://www.opengis.net/def/crs/OGC/1.3/CRS84` -> `CRS84`. Or are SRIDs external to OGC and so there is no mapping? I had a look and found some resources in XML but not sure I've found the right info.

recalcitrantsupplant replied:
The logic has been updated to extract the last part of URLs supplied, send these to geomet's dumps function (the assumption being geomet will correctly invert the axes order if required), but then return the full CRS URL originally supplied always. https://github.com/RDFLib/prez/pull/373/commits/be7c9e04fbe467c6de346586e98d2dc92e35841c 

There's obviously cases currently where the axes order should/shouldn't be inverted that geomet doesn't cover, if the reference data is supplied it will be easier to implement logic for this.

recalcitrantsupplant commented on prez/services/query_generation/spatial_filter.py:
```diff
@@ -0,0 +1,358 @@
+
+def get_wkt_from_coords(coordinates, geom_type: str, filter_crs):
+    max_decimals = find_max_decimals([(geom_type, coordinates, None)])
+    if filter_crs == "http://www.opengis.net/def/crs/OGC/1.3/CRS84":
```
I'll see what the response is to the comment below, and depending on that I might map all of the URIs including urn forms to SRIDs. Otherwise I can just add these in too yes.

recalcitrantsupplant replied:
Have implemented function to take last part of CRS URL so that the PR can be merged. https://github.com/RDFLib/prez/pull/373/commits/be7c9e04fbe467c6de346586e98d2dc92e35841c

lalewis1 reviewed (approved):
thanks for all the work here David.

LGTM

ashleysommer reviewed (approved):
👍Lets merge it!
